src/storage-client/src/types/sources.rs

```diff
@@ -1299,9 +1309,36 @@ impl UnplannedSourceEnvelope {
                         (key_desc.with_key(vec![0]).concat(value_desc), Some(vec![0]))
                     }
                 };
+
+                // Getting order by column index from the metadata
+                let order_by_index = match order_by {
+                    Some(order_col) => {
+                        match metadata_desc.get_by_name(&order_col.clone().into()) {
+                            Some((position, column_type)) => {
+                                match column_type.scalar_type {
+                                    ScalarType::Timestamp => {
+                                        Ok::<std::option::Option<usize>, anyhow::Error>(Some(
+                                            // getting absolute position in the row as metadata is always appended to keyed
+                                            position + keyed.arity(),
+                                        ))
+                                    }
+                                    _ => bail!("Only TIMESTAMP is supported as an order by column"),
+                                }
+                            }
+                            None => bail!("Could not find {} in included metadata", order_col),
+                        }
+                    }
+                    None => Ok(None),
+                }?;
```

Comment:
Recommend that you hoist this logic into the planner here:

https://github.com/MaterializeInc/materialize/blob/c7b89d67aa476929b5db51905e3b3f7b8845b58d/src/sql/src/plan/statement/ddl.rs#L910

In particular you'll be able to check that the named column is literally the Kafka timestamp column that the user has requested with `INCLUDE TIMESTAMP`, vs any metadata column with type timestamp. (At present, only Kafka sources have metadata columns of type timestamp, and only one such column, but that might not always be the case!)

This dovetails with the previous comment about changing to `UnplannedEnvelope::Upsert { order_by_column }`; you'll just be able to plumb the `order_by_index` straight into the `SourceEnvelope`.

Reply:
I will update it to be part of the upsert enum. 

Since I was parsing the column name as an identifier, the user could have done `INCLUDE TIMESTAMP as ts` and then use `ORDER BY ts`. 

OOC in case say another timestamp based metadata column gets added to kafka, would it be too bad if that's also allowed?

Nevertheless I think I can make sure that only the particular `TIMESTAMP` included column is referred.

Reply:
So, I was trying to move this in ddl.rs. But I would still need access to `keyed.arity()` to get the absolute index in the entire row (key + value + metadata), which is only available here?

Reply:
I refactored it out. But later still having to convert it to absolute row position from metadata position [code](https://github.com/MaterializeInc/materialize/pull/18567/files#diff-3fab9304494d4712cbc38dceda54af0815ec689eb37b17b6c60d193f7b52f2c8R1247-R1248)